# Procesamiento de datos con Python Santander 2022

Módulo 2
Post6 (APi/Webscraping) 

Equipo 9. Integrantes:

* Isidro Amaro
* Francisco Gómez
* Tomás Antonio Hernández Pineda
* José Luis Herrera Gallardo
* Christian E. Millán Hernández

## Documentos
   - [Colab Jupyter NoteBook:](https://colab.research.google.com/drive/16cZDxkxTPmwky0vxjkXb-lEh-Cb8Xhyt?usp=share_link)

   - [Web Scraping Colab](https://colab.research.google.com/drive/18mS4nu7Ed7FUquBDkZl1K-HB2-nvYxfK?usp=share_link)

   - [Gdrive](https://drive.google.com/drive/folders/1JMUBfeMb21vYTT8YQgvkbrksdHtsgN_A?usp=share_link)

- [Github repo](https://github.com/BeduDSEquipo9/G2DSF2_Python)

## Contenido


* [Obtención de datos adicionales con WebScraping en lugar de API](https://colab.research.google.com/drive/18mS4nu7Ed7FUquBDkZl1K-HB2-nvYxfK#scrollTo=PWnz4kaGMozP&line=1&uniqifier=1)
* [Exploración de los datos](https://colab.research.google.com/drive/18mS4nu7Ed7FUquBDkZl1K-HB2-nvYxfK#scrollTo=JzPKfkUm48Xf&line=2&uniqifier=1)
   - Primera de exploración d datos
* Limpieza de los datos y procesamiento de datos
* [Extrayendo datos agregados por Estados y Ciudades](https://colab.research.google.com/drive/18mS4nu7Ed7FUquBDkZl1K-HB2-nvYxfK#scrollTo=F78qYdLOrR5F&line=1&uniqifier=1)
  - [Contestando algunos preguntas con nuestro Dataset limpio](https://colab.research.google.com/drive/18mS4nu7Ed7FUquBDkZl1K-HB2-nvYxfK#scrollTo=uYSd9vCzU9TN&line=1&uniqifier=1)






In [ ]:
#Importando librerias
import pandas as pd
import numpy as np

El conjunto de datos se encuentra almacenado en Google Drive y Github. 
<p>Si desea utilizar Gdrive, ejecute el siguente comando,
<p> para habilitar el acceso desde Google Colab. 
<p>(Además de cargar, se podrá guardar los datasets procesados.)

In [ ]:
#Se requiere si los Datasets preocesados se quieren guardar en Gdrive a Github
#Se puede deshabilitar u omitir si los datasets a usar son de Github.
#Existe una copia del Dataset Airbnb en github asi como de los extraidos por Webscraping

from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


El conjunto de datos esta almacenada en formato `csv`, por lo que se utiliza `pd.read_csv()` para crear un objeto de tipo DataFrame.

In [ ]:
#NOTAS GDRIVE
#DATASET de AIRBNB Inicial guardado en GDrive
#file_path='/content/drive/MyDrive/BEDUDSF2M3Python2023Proy/Proy/Datasets/airbnb.csv'

#NOTAS para Github
#Ruta al repo del archivo Github
#Github: https://github.com/BeduDSEquipo9/G2DSF2_Python/tree/Developer 
#redireccionar a la rama main o si se estan realizando evaluaciones con datasets ubicados en otras ramas
#Ejemplo de Datset ubicado en rama Developer
#file_path = 'https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/Developer/Datasets/airbnb.csv'

#REdireccionando a rama main del repo.
file_path = 'https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/main/Datasets/airbnb.csv'
data=pd.read_csv(file_path)

# Postwork 6

## Obtención de datos adicionales con WebScraping en lugar de API


  Airbnb no brinda acceso a su API a terceros debido a la privacidad de la información de sus usuarios y la protección de sus datos. Para nuestro proyecto encontramos que las APIs de AirBnb se encuentran cerradas a menos que seas dueño de una propiedad en la plataforma o un proveedor de servicios para mantenimiento y administración de inmuebles. 

  Durante este proyecto nos surgio otra, como se compara una renta de corto plazo vs una renta de largo plazo. Vamos a comparar el valor de la renta de corto plazo de Airbnb, obtenido en el modelo predictor vs el valor de la renta mensual. Para lo cual exploramos otras API de empresas de anuncios de renta de inmuebles. Entre ellas nos centramos en dos API sin costo, pero el proceso para la liberación de los accesos toma tiempo. Una de ellas es [Zillow](www.zillow.com), el portal de inmuebles más reconocido en el mercado estadounidense. Para analisis futuros podriamos utilizar los datos de [Inmuebles24.com](Inmuebles24.com), un homólogo de Zillow en nuestro país.

  En lugar de explotar un API, decidimos realizar un barrido de datos Webscraping del portal Zillow.com con python. Esta estrategia permite tener acceso a los datos necesarios para la comparación sin tener que depender de una API externa y, por lo tanto, puede ser una solución más asequible y controlable.

  El codigo de Webscraping se encuentra en nuestro [repo](https://github.com/BeduDSEquipo9/G2DSF2_Python/tree/main/Datasets) en Github. *((No lo agregamos a este Colab, por los problemas que pueden ocurrir por ip-Blocking cuando AWS proveedor de los servicios de Nube para Zillow, se da cuenta que estamos haciendo un barrido de su pagina.))*
  
  Como ya observamos en el [Analisis exploratorio](https://colab.research.google.com/drive/16cZDxkxTPmwky0vxjkXb-lEh-Cb8Xhyt#scrollTo=xH1c2lWNQsLx&line=1&uniqifier=1) el Dataset de AirBnb corresponde a 6 ciudades de de Estados Unidos: 
  
  








In [ ]:
data['city'].unique()

array(['NYC', 'LA', 'Chicago', 'Boston', 'SF', 'DC'], dtype=object)

[Script de WebScraping](https://github.com/BeduDSEquipo9/G2DSF2_Python/tree/main/Post6_Webscrapping/Zillowv2.py)

```
#Lista de Ciudades, obtenidas del script de Webscraping
#city = ["nyc","Los-Angeles","chicago","boston","San-Francisco","washington-dc"]
```



 
  Desarrollamos varias versiones del código para la extracción de datos de las mismas 6 ciudades. Empezamos haciendo una análisis visual del portal de Zillow encontramos varias formas de obtener resultados. (Si no puedes accesar ve a nuestro [repo](https://github.com/BeduDSEquipo9/G2DSF2_Python/tree/main/Datasets) en Github.
  
 - ![Zillow00.png](https://github.com/BeduDSEquipo9/G2DSF2_Python/blob/main/Post6_Webscrapping/images/Zillow00.png)
 
 - ![Zillow01.png](https://github.com/BeduDSEquipo9/G2DSF2_Python/blob/main/Post6_Webscrapping/images/Zillow00.png)

  Podemos emplear el bloque de búsqueda (SearchBox) en la página principal o desde las páginas secundarias. La petición enviada puede ser del tipo GET y un Header muy complejo que incluye datos de posición de un polígono en un mapa o un URL al nombre de una ciudad en específica. 

- [https://www.zillow.com/homes/Los-Angeles,-CA_rb/](https://www.zillow.com/homes/Los-Angeles,-CA_rb/)

- [https://www.zillow.com/homes/for_rent/Los-Angeles/](https://www.zillow.com/homes/for_rent/Los-Angeles/)


  Cuando exploramos estos enlaces directamente, observamos en la parte derecha una lista de dos columnas con un total de 40 propiedades por página y sus datos básicos, a la izquierda un mapa de la ciudad con marcadores indicando espacialmente la ubicación de las propiedades listadas en el sistema. Al llegar a la parte inferior existe paginación de los datos restantes.
Zillow es un sitio con carga dinámica a través de json y varios HTML, CSS y JS. Para hacer un barrido o Web Scraping del Sitio de Zillow podemos emplear las librerías Requests, Beautifulsoup y Selenium juntas o separadas.  
	
- [Requests:](https://requests.readthedocs.io/en/latest/)
  - En el sistema: pip install requests
  + En python: import requests
- [Beautifulsoup:](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
  - En el sistema: pip install beautifulsoup4
  - En el sistema: pip install lxml (un parser para xml y html)
  - En python: import beautifulsoup4 as bs4
- [Selenium:	](https://selenium-python.readthedocs.io/)
	- En el sistema: pip install selenium
	- En python: from selenium import webdriver

Encontramos en Zillow regresa un JSON con el codigo de 40 propiedades.

Si bien una forma muy elegante seria utilizando Selenium Driver, el empleo de requests contra el  url1 = 'https://www.zillow.com/homes/for_sale/'+city y la busqueda del JSON cat1 con elemento searchResults con subelemento listResults es más rápido.



In [ ]:
#Abre un archivo1
file_path0='https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/main/Post6_Webscrapping/WebScrapping_Datasets/data_NYC.csv'
df=pd.read_csv(file_path0)

#Abre un archivo2
file_path1='https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/main/Post6_Webscrapping/WebScrapping_Datasets/data_Boston.csv'
df_tempo=pd.read_csv(file_path1);
df=df.append(df_tempo,ignore_index=True)

#Abre un archivo3
file_path2='https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/main/Post6_Webscrapping/WebScrapping_Datasets/data_washigton-dc.csv'
df_tempo=pd.read_csv(file_path2);
df=df.append(df_tempo,ignore_index=True)

#Abre un archivo4
file_path3='https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/main/Post6_Webscrapping/WebScrapping_Datasets/data_San-Francisco.csv'
df_tempo=pd.read_csv(file_path3);
df=df.append(df_tempo,ignore_index=True)

#Abre un archivo5
file_path4='https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/main/Post6_Webscrapping/WebScrapping_Datasets/data_Los-Angeles.csv'
df_tempo=pd.read_csv(file_path4);
df=df.append(df_tempo,ignore_index=True)




In [ ]:
#Guarda los diferentes Datasets en 1 solo CSV en GDRIVE 
df.to_csv('/content/drive/MyDrive/BEDUDSF2M3Python2023Proy/Proy/Datasets/MergedData.csv')

In [ ]:
#Usa el archivo en Gdrive o la versión guardada en el repo
#df = pd.read_csv('/content/drive/MyDrive/BEDUDSF2M3Python2023Proy/Proy/Datasets/MergedData.csv')

#Usando una version en 
#file_path='https://raw.githubusercontent.com/BeduDSEquipo9/G2DSF2_Python/main/Post6_Webscrapping/WebScrapping_Datasets/MergedData.csv'
#df=pd.read_csv(file_path);



##Exploración de los datos
 - ### Primera exploración de los datos de Webscraping

In [ ]:
# Explore Dataset
print('shape:', df.shape)

In [ ]:
df.head(10)

In [ ]:
df.info()

In [ ]:
#Hay nulos?
df.isnull().values.any()

True

In [ ]:
#Dame una rediografia del DataFrame de datos nulos por columas.
df.isnull().sum()

In [ ]:
#¿Cuál es el porcentaje de nulos?
nulos_porcentaje=(df.isnull().sum()/len(df))*100 #Porcentaje de valores nulos
nulos_porcentaje.sort_values(ascending= False)

##Revisión aleatoria de registros

---



In [ ]:
df.iloc[1, :]

## Limpieza de los datos de Webscraping

###Eliminando columnas
Observemos cuales columnas vamos a eliminar puesto que no sirven para el analisis.

In [ ]:
print("columnas", df.columns)
lista_columnas = len(df.columns)
print("Numero de columnas:", lista_columnas)

columnas Index(['zpid', 'id', 'lotId', 'imgSrc', 'hasImage', 'statusType', 'statusText',
       'detailUrl', 'latLong', 'units', 'variableData', 'badgeInfo',
       'buildingName', 'isBuilding', 'address', 'addressStreet',
       'addressState', 'addressCity', 'addressZipcode', 'providerListingId',
       'canSaveBuilding', 'has3DModel', 'isFeaturedListing', 'isSaved', 'list',
       'relaxed', 'streetViewMetadataURL', 'streetViewURL', 'area',
       'availabilityDate', 'baths', 'beds', 'countryCurrency',
       'hasAdditionalAttributions', 'hasVideo', 'hdpData', 'isHomeRec',
       'isUndisclosedAddress', 'isUserClaimingOwner', 'isUserConfirmedClaim',
       'isZillowOwned', 'pgapt', 'price', 'sgapt',
       'shouldShowZestimateAsPrice', 'unformattedPrice', 'zestimate',
       'brokerName', 'hasOpenHouse', 'openHouseDescription',
       'openHouseEndDate', 'openHouseStartDate', 'carouselPhotos'],
      dtype='object')
Numero de columnas: 53


Preparamos la lista de columnas a eliminar y revisamos el resultado.
<p>
El Dataset contempla 11 columnas que aun deben ser procesadas 

In [ ]:
# REMOVE UNUSED COLUMNS FROM 54 kept X columns
df_dropped = df.drop(columns=['zpid', 'imgSrc', 'hasImage', 'statusType',
                              'statusText', 'detailUrl', 'latLong', 'variableData',
                              'badgeInfo', 'buildingName', 'isBuilding', 'address', 'addressStreet',
                              'providerListingId',
                              'canSaveBuilding', 'has3DModel', 'isFeaturedListing', 'isSaved', 'list',
                              'relaxed', 'streetViewMetadataURL', 'streetViewURL', 'area',
                              'availabilityDate', 'countryCurrency',
                              'hasAdditionalAttributions', 'hasVideo', 'hdpData', 'isHomeRec',
                              'isUndisclosedAddress', 'isUserClaimingOwner', 'isUserConfirmedClaim',
                              'isZillowOwned', 'pgapt', 'sgapt',
                              'shouldShowZestimateAsPrice', 'zestimate',
                              'brokerName', 'hasOpenHouse', 'openHouseDescription',
                              'openHouseEndDate', 'openHouseStartDate', 'carouselPhotos'])

print("Lista de columnas que tiene nuestro Dataset: ", df_dropped.columns)
lista_columnas = len(df_dropped.columns)
print("Numero de columnas que mantendremos :", lista_columnas)

Lista de columnas que tiene nuestro Dataset:  Index(['id', 'lotId', 'units', 'addressState', 'addressCity', 'addressZipcode',
       'baths', 'beds', 'price', 'unformattedPrice'],
      dtype='object')
Numero de columnas que mantendremos : 10


## Segunda exploración de los datos de Webscraping
Exploremos el Dataset restante despues de eliminar las columnas no esenciales.


In [ ]:
df_dropped.describe()

In [ ]:
df_dropped.head(10)

In [ ]:
df_dropped.info()

##Nuevamente revisión aleatoria de registros


In [ ]:
df_dropped.iloc[1, :]

Observese la columna 'units' trae arreglo de Directorios (JSONs)
<p> Cada elemento pertenece a una propiedad en la misma dirección o predio.

In [ ]:
df_dropped['units']

0       [{'price': '$3,100+', 'beds': '0'}, {'price': ...
1       [{'price': '$2,250+', 'beds': '0'}, {'price': ...
2       [{'price': '$3,664+', 'beds': '0'}, {'price': ...
3       [{'price': '$5,075+', 'beds': '1'}, {'price': ...
4       [{'price': '$3,411+', 'beds': '0'}, {'price': ...
                              ...                        
1995    [{'price': '$1,960+', 'beds': '0'}, {'price': ...
1996    [{'price': '$2,175+', 'beds': '0'}, {'price': ...
1997    [{'price': '$2,133+', 'beds': '1'}, {'price': ...
1998    [{'price': '$2,784+', 'beds': '1'}, {'price': ...
1999                                                  NaN
Name: units, Length: 2000, dtype: object

### Veamos cuantos datos faltan en el Dataset.

In [ ]:
df_dropped.isnull().values.any()

True

In [ ]:
# CHECK MISSING DATA
df_dropped.isna().sum()


In [ ]:
nulos_porcentaje=(df_dropped.isnull().sum()/len(df))*100 #Porcentaje de valores nulos
nulos_porcentaje.sort_values(ascending= False)

###Ejemplos de datos en la columna 'units'


```
[{'price': '$3,664+', 'beds': '0'}, {'price': '$4,000+', 'beds': '1'}, {'price': '$5,630+', 'beds': '2'}, {'price': '$7,108+', 'beds': '3'}]
```
Este arreglo de diccionarios o Jsons representa que en la misma direccion existen 4 propiedades diferentes en renta. 

Para lo cual podemos hacer varias sustituciones con para tener separadores especiales. Uno para separa en renglones y otro en columnas

Buscamos algo parecido :
```
[{'price': '$3,664+', 'beds': '0'}
 {'price': '$4,000+', 'beds': '1'}
  {'price': '$5,630+', 'beds': '2'}
   {'price': '$7,108+', 'beds': '3'}]
```
<p>Vamos a todos los siguientes caracteres(innecesarios):"[{'+}]"

<p>Iniciemos separarando los datos con el comando explode de Pandas. 

<p>En este caso para convertir una columna necesitamos identificar un separador. 
<p>Por lo cual empleamos el comando split primero para indicar una cadena separadora: '}, {' 
<p>Posteriormente llamamos a explode para separar en renglones los datos.
<p>Se observaŕa el indice se repite, junto con todos los datos en las columnas restantes del registro.

In [ ]:
#Dejamos el DataFrame sin columnas y pasamos a REv1 (R1)
df_R1 = df_dropped

#Indiquemos el caracter o patron separador de los datos en la columna 'units'
df_R1['units'] = df_R1['units'].str.split('}, {')

#Ejecuta explode de PAndas para separar las listas de la columna en renglores
df_R1 =df_R1.explode('units').reset_index(drop=True)



Observamos que pasamos de 2000 registros a 3675

In [ ]:
#Como queda después de la transformación 'units'
#Esta celda solo se ejecuta si se quiere ver el resultado de la 
#transformacion sobre algunas columnas.
#Se puede sustituir por .head() o .tail(), por comodidad dejamos asi.
#df_R1['units']
df_R1

id         lotId  \
0       40.70457_rrc2*-rrc_74.00886  1.044499e+09   
1       40.70457_rrc2*-rrc_74.00886  1.044499e+09   
2       40.70457_rrc2*-rrc_74.00886  1.044499e+09   
3      40.588963_rrc2*-rrc_73.79774  2.355012e+09   
4      40.588963_rrc2*-rrc_73.79774  2.355012e+09   
...                             ...           ...   
3670   34.05514_rrc2*-rrc_118.30745  1.001581e+09   
3671   34.05514_rrc2*-rrc_118.30745  1.001581e+09   
3672  33.955074_rrc2*-rrc_118.44068  1.001554e+09   
3673  33.955074_rrc2*-rrc_118.44068  1.001554e+09   
3674                     2069756690           NaN   

                                  units addressState    addressCity  \
0     [{'price': '$3,100+', 'beds': '0'           NY       New York   
1       'price': '$3,921+', 'beds': '1'           NY       New York   
2     'price': '$6,395+', 'beds': '2'}]           NY       New York   
3     [{'price': '$2,250+', 'beds': '0'           NY        Arverne   
4     'price': '$2,800+', 'beds': '1'}]           NY        Arverne   
...                                 ...          ...            ...   
3670  [{'price': '$2,133+', 'beds': '1'           CA    Los Angeles   
3671  'price': '$2,831+', 'beds': '2'}]           CA    Los Angeles   
3672  [{'price': '$2,784+', 'beds': '1'           CA  Playa Del Rey   
3673  'price': '$3,380+', 'beds': '2'}]           CA  Playa Del Rey   
3674                                NaN           CA  Panorama City   

      addressZipcode  baths  beds      price  unformattedPrice  
0              10005    NaN   NaN        NaN               NaN  
1              10005    NaN   NaN        NaN               NaN  
2              10005    NaN   NaN        NaN               NaN  
3              11692    NaN   NaN        NaN               NaN  
4              11692    NaN   NaN        NaN               NaN  
...              ...    ...   ...        ...               ...  
3670           90006    NaN   NaN        NaN               NaN  
3671           90006    NaN   NaN        NaN               NaN  
3672           90293    NaN   NaN        NaN               NaN  
3673           90293    NaN   NaN        NaN               NaN  
3674           91402    2.0   2.0  $2,595/mo            2595.0  

[3675 rows x 10 columns]

## Eliminamos todos los siguientes caracteres(innecesarios):"[{'+}]" con su codigo hexadecimal.


In [ ]:
#Eliminamos todos los siguientes caracteres(innecesarios):"[{'+}]" con su codigo hexadecimal.
df_R1["units"]=df_R1["units"].str.replace(r'[\x7B\x7D\x5B\x5D\+ \x27price\x27\x3A]', '')

<ipython-input-27-722a10d46d4e>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  df_R1["units"]=df_R1["units"].str.replace(r'[\x7B\x7D\x5B\x5D\+ \x27price\x27\x3A]', '')


In [ ]:
#Como queda después de la transformación
#Esta celda solo se ejecuta si se quiere ver el resultado de la 
#transformacion sobre algunas columnas.
#Se puede sustituir por df_R1 con o sin .head() o .tail(), por comodidad dejamos asi.
df_R1.head(10)

id         lotId        units addressState  \
0    40.70457_rrc2*-rrc_74.00886  1.044499e+09  $3,100,bds0           NY   
1    40.70457_rrc2*-rrc_74.00886  1.044499e+09  $3,921,bds1           NY   
2    40.70457_rrc2*-rrc_74.00886  1.044499e+09  $6,395,bds2           NY   
3   40.588963_rrc2*-rrc_73.79774  2.355012e+09  $2,250,bds0           NY   
4   40.588963_rrc2*-rrc_73.79774  2.355012e+09  $2,800,bds1           NY   
5  40.750156_rrc2*-rrc_73.936195  2.560384e+09  $3,664,bds0           NY   
6  40.750156_rrc2*-rrc_73.936195  2.560384e+09  $4,000,bds1           NY   
7  40.750156_rrc2*-rrc_73.936195  2.560384e+09  $5,630,bds2           NY   
8  40.750156_rrc2*-rrc_73.936195  2.560384e+09  $7,108,bds3           NY   
9  40.691814_rrc2*-rrc_73.982994  2.009532e+09  $5,075,bds1           NY   

        addressCity  addressZipcode  baths  beds price  unformattedPrice  
0          New York           10005    NaN   NaN   NaN               NaN  
1          New York           10005    NaN   NaN   NaN               NaN  
2          New York           10005    NaN   NaN   NaN               NaN  
3           Arverne           11692    NaN   NaN   NaN               NaN  
4           Arverne           11692    NaN   NaN   NaN               NaN  
5  Long Island City           11101    NaN   NaN   NaN               NaN  
6  Long Island City           11101    NaN   NaN   NaN               NaN  
7  Long Island City           11101    NaN   NaN   NaN               NaN  
8  Long Island City           11101    NaN   NaN   NaN               NaN  
9          Brooklyn           11201    NaN   NaN   NaN               NaN

Separamos la columna'units'en 2 columnas y depositamos los datos 
<p>sobre las columnas 'beds' y 'unformattedPrice"(precio sin formato) 
<p>que ya teniamos. 
<p>(De esta manera reducimos la cantidad de datos faltantes "NaN")

In [ ]:
#Este comando se usa si se ejecutan los comandos desde python terminal.
#df_R1['unformattedPrice']=(df_R1['units'].str.split(',bds', expand=True))[0]
#df_R1['beds']=(df_R1['units'].str.split(',bds', expandmain/Post6_Webscrapping/WebScrapping_Datasets/=True))[1]

#separando en columnas, este comando solo corre en Colab, ejecutando el comando 
#como script.py en terminal no corre el programa.Comentese y sustituya por las 2
#líneas comentadas al inicio de está celda
df_R1[['unformattedPrice', 'beds']] = df_R1['units'].str.split(',bds', expand=True)

##Eliminemos el formato a la columna 'price'

Con el objeto de hacer operaciones de agregación como datos de tipo número posteriormente.
<BR>
(Recordar que '\D'= pattern digito(deja solo los números)

In [ ]:
from pandas.core.reshape.merge import string
#Eliminado todo lo que no sea numero.
df_R1["price"]=df_R1["price"].str.replace(r'\D', '')

<ipython-input-30-5a1aba46b252>:3: FutureWarning: The default value of regex will change from True to False in a future version.
  df_R1["price"]=df_R1["price"].str.replace(r'\D', '')


#####Revisemos los cambios en la columna 'price'
Observamos que los signos de "$" y "/mo" se eliminan puesto que \D deja solo la cantidad numerica. 
Despues aplicamos to:numeric para adegurar el dato es de tipo numero par apoder hacer calgulos con funciones de aggregacion al final del procesamiento.



In [ ]:
#Como queda después de la transformación
#Esta celda solo se ejecuta si se quiere ver el resultado de la 
#transformacion sobre algunas columnas.
#Se puede sustituir por df_R1 con o sin .head() o .tail(), por comodidad dejamos asi.
df_R1

id         lotId        units addressState  \
0       40.70457_rrc2*-rrc_74.00886  1.044499e+09  $3,100,bds0           NY   
1       40.70457_rrc2*-rrc_74.00886  1.044499e+09  $3,921,bds1           NY   
2       40.70457_rrc2*-rrc_74.00886  1.044499e+09  $6,395,bds2           NY   
3      40.588963_rrc2*-rrc_73.79774  2.355012e+09  $2,250,bds0           NY   
4      40.588963_rrc2*-rrc_73.79774  2.355012e+09  $2,800,bds1           NY   
...                             ...           ...          ...          ...   
3670   34.05514_rrc2*-rrc_118.30745  1.001581e+09  $2,133,bds1           CA   
3671   34.05514_rrc2*-rrc_118.30745  1.001581e+09  $2,831,bds2           CA   
3672  33.955074_rrc2*-rrc_118.44068  1.001554e+09  $2,784,bds1           CA   
3673  33.955074_rrc2*-rrc_118.44068  1.001554e+09  $3,380,bds2           CA   
3674                     2069756690           NaN          NaN           CA   

        addressCity  addressZipcode  baths beds price unformattedPrice  
0          New York           10005    NaN    0   NaN           $3,100  
1          New York           10005    NaN    1   NaN           $3,921  
2          New York           10005    NaN    2   NaN           $6,395  
3           Arverne           11692    NaN    0   NaN           $2,250  
4           Arverne           11692    NaN    1   NaN           $2,800  
...             ...             ...    ...  ...   ...              ...  
3670    Los Angeles           90006    NaN    1   NaN           $2,133  
3671    Los Angeles           90006    NaN    2   NaN           $2,831  
3672  Playa Del Rey           90293    NaN    1   NaN           $2,784  
3673  Playa Del Rey           90293    NaN    2   NaN           $3,380  
3674  Panorama City           91402    2.0  NaN  2595              NaN  

[3675 rows x 10 columns]

In [ ]:
df_R1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3675 entries, 0 to 3674
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                3675 non-null   object 
 1   lotId             2925 non-null   float64
 2   units             2981 non-null   object 
 3   addressState      3675 non-null   object 
 4   addressCity       3675 non-null   object 
 5   addressZipcode    3675 non-null   int64  
 6   baths             694 non-null    float64
 7   beds              2981 non-null   object 
 8   price             694 non-null    object 
 9   unformattedPrice  2981 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 287.2+ KB


## Removemos el formato de moneda ala columna 'unformattedPrice'
y completamos los registros NaN con los valores de la columna 'price'

In [ ]:
def double_to_money(value):

    return f'${value}'
#Comentamos lallamada a esta función puesto que al sacar 
#df_R1['price']=df_R1['price'].map(double_to_money)

df_R1['unformattedPrice'] = df_R1['unformattedPrice'].fillna(df_R1['price'])
df_R1['unformattedPrice'] = df_R1['unformattedPrice'].str.replace(r'[\x24\x2C]', '')
df_R1['unformattedPrice'] = pd.to_numeric(df_R1['unformattedPrice'])


<ipython-input-34-b4d2170421a6>:8: FutureWarning: The default value of regex will change from True to False in a future version.
  df_R1['unformattedPrice'] = df_R1['unformattedPrice'].str.replace(r'[\x24\x2C]', '')


In [ ]:
#Como queda después de la transformación
#Esta celda solo se ejecuta si se quiere ver el resultado de la 
#transformacion sobre algunas columnas.
#Se puede sustituir por df_R1 con o sin .head() o .tail(), por comodidad dejamos asi.
df_R1

## Rellenando y eliminando más columnas   

Al inicio observamos que este Dataset tenia 3 columna de IDs:
- zpId
- lotId
- Id

Ya eliminamos zpId, vamos a completar la columna 'lotid' con los valores de 'Id'.

Posterioremente eliminaremos las columnas que ya dejaron de ser utilies como:
- 'units'
- 'price'
- 'id'

Vamos a renombrar las columnas del Dataframe por unos mas significativos.

In [ ]:
#Rellenar columna 'id' con los valores de 'lotId'.
df_R1['lotId'] = df_R1['lotId'].fillna(df_R1['id'])

#eliminaremos las columnas que ya dejaron de ser utiles 
#df_R1=df_R1.drop(columns=['Unnamed: 0', 'units','price','id'])
df_R1=df_R1.drop(columns=['units','price','id'])

#Reset al index.
df_R1=df_R1.reset_index(drop=True)

#Preparando lista de renombrado de columnas
#Recuerda izq. 'valor actual': 'der. nuevovalor '
column_name_mapping={
    #'lotId':'LotId',
    'addressState':'state',
    'addressCity':'city',
    'addressZipcode':'zipcode',
    #'baths':'Baths',
    #'beds':'Beds',
    'unformattedPrice':'price'
}

#Ejecutamos renombrado de columnas.
df_R1 = df_R1.rename(columns=column_name_mapping)

In [ ]:
#Como queda después de la transformación
#Esta celda solo se ejecuta si se quiere ver el resultado de la 
#transformacion sobre algunas columnas.
#Se puede sustituir por df_R1 con o sin .head() o .tail(), por comodidad dejamos asi.
df_R1


En el caso de la columna baños ('baths') se realizará el reemplazo de los `NaN` en el supuesto de que la propiedad al menos debe tener un baño.

De forma siliar realizaremos el reemplazo de `NaN` en el resto de las columnas que presentan ausencia de datos: `beds`

In [75]:
df_R1['baths'].fillna(0, inplace=True)
df_R1['baths'].replace(0, 1., inplace=True)
df_R1['beds'].fillna(0, inplace=True)
df_R1['beds'].replace(0, 1., inplace=True)

In [76]:
#Como queda después de la transformación
#Esta celda solo se ejecuta si se quiere ver el resultado de la 
#transformacion sobre algunas columnas.
#Se puede sustituir por df_R1 con o sin .head() o .tail(), por comodidad dejamos asi.
df_R1

lotId state           city zipcode  baths  beds   price
0     1044499437.0    NY       New York   10005      1     1  3100.0
1     1044499437.0    NY       New York   10005      1     1  3921.0
2     1044499437.0    NY       New York   10005      1     2  6395.0
3     2355011519.0    NY        Arverne   11692      1     1  2250.0
4     2355011519.0    NY        Arverne   11692      1     1  2800.0
...            ...   ...            ...     ...    ...   ...     ...
3670  1001580711.0    CA    Los Angeles   90006      1     1  2133.0
3671  1001580711.0    CA    Los Angeles   90006      1     2  2831.0
3672  1001553584.0    CA  Playa Del Rey   90293      1     1  2784.0
3673  1001553584.0    CA  Playa Del Rey   90293      1     2  3380.0
3674    2069756690    CA  Panorama City   91402      2     1  2595.0

[3675 rows x 7 columns]

In [77]:
df_R1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3675 entries, 0 to 3674
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   lotId    3675 non-null   object 
 1   state    3675 non-null   object 
 2   city     3675 non-null   object 
 3   zipcode  3675 non-null   object 
 4   baths    3675 non-null   int64  
 5   beds     3675 non-null   int64  
 6   price    3675 non-null   float64
dtypes: float64(1), int64(2), object(4)
memory usage: 201.1+ KB


Hagamos castigs de tipos de datos.

In [78]:
from pandas.core.reshape.merge import string

#Casting a tipo de datos
#df_R1['price']=pd.to_numeric(df_R1['price'])

df_R1['baths']=df_R1['baths'].astype(int)
df_R1['beds']=df_R1['beds'].astype(int)
df_R1['price']=df_R1['price'].astype(float)
df_R1['zipcode']=df_R1['zipcode'].astype(str)

In [79]:
df_R1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3675 entries, 0 to 3674
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   lotId    3675 non-null   object 
 1   state    3675 non-null   object 
 2   city     3675 non-null   object 
 3   zipcode  3675 non-null   object 
 4   baths    3675 non-null   int64  
 5   beds     3675 non-null   int64  
 6   price    3675 non-null   float64
dtypes: float64(1), int64(2), object(4)
memory usage: 201.1+ KB


##Revisemos nuestro avance hasta el momento

In [ ]:
print('shape:', df_R1.shape)
print("describe", df_R1.describe())
print("head", df_R1.head(10))
print("Info:", df_R1.info())
print("columnas", df_R1.columns)
lista_columnas = len(df_R1.columns)
print("Numero de columnas:", lista_columnas)

___

## Confirmemos el estado del Dataset
 (En relación al número de NaNs que aun persisten.)


In [81]:
#Dame una rediografia del DataFrame de datos nulos por columas.
#¿Cuántos valores nulos quedan?
df_R1.isnull().sum()


lotId      0
state      0
city       0
zipcode    0
baths      0
beds       0
price      0
dtype: int64

- Se observa que pasamos de 2000 registros a 3675,  de los cuales 694<p> no tienen precio ni numero de habitaciones. 
<p>
- <p>Suponemos son las propiedades que solo se anuncian y <p>para conocer más información debemos contactar al agente inmobiliario.
- <p>No eliminaremos por el momento los registros y aśi conocer donde hay mas opciones de renta. 


In [83]:
#Generamos una copia del DataFrame procesado y lo denominamos limpio
df_clean=df_R1


## Guardando csv procesado en Gdrive

In [84]:
#Usa el archivo en Gdrive o la versión guardada en el repo
df_R1.to_csv('/content/drive/MyDrive/BEDUDSF2M3Python2023Proy/Proy/Datasets/ProcessedData.csv')

___

In [85]:
print("Tamaño:",df_R1.groupby('state').size())

Tamaño: state
CA    1400
DC     891
MA     756
NY     628
dtype: int64


##Conclusión

In [86]:
print("\nCuenta Valores1: ",df_R1.groupby('state')['city'].value_counts())


Cuenta Valores1:  state  city               
CA     San Francisco          658
       Los Angeles            499
       North Hollywood         38
       Canoga Park             28
       Woodland Hills          26
                             ... 
NY     Rego Park                2
       Ridgewood                2
       Forest Hills             1
       Jackson Heights          1
       Springfield Gardens      1
Name: city, Length: 66, dtype: int64


In [ ]:
print("\nCuenta Valores2: ",df_R1.groupby(['state','city'])['zipcode'].value_counts())


Cuenta Valores2:  state  city           zipcode
CA     Canoga Park    91303      15
                      91304      13
       Chatsworth     91311       3
       Culver City    90232       5
       Encino         91316      10
                                 ..
NY     Staten Island  10304       4
                      10301       3
                      10310       1
                      10314       1
       Woodside       11377       4
Name: zipcode, Length: 281, dtype: int64


###Contestando algunos preguntas con nuestro Dataset limpio
Para contestar algunas preguntas, en esta sección se retoman algunos de los planteamos realizados en la descripción del problema:

1. ¿Cuál es costo promedio de una propiedad en airbnb?
2. ¿Cuál es la ciudad que tienen el costo promedio más caro?
3. ¿Cuál es el costo promedio de una renta de acuerdo al tipo de propiedad?
4. ¿Cuales son los vecindarios más costosos y en que ciudad se encuentran?

**Pregunta 1. ¿Cuál es costo promedio de una propiedad en airbnb?**

Con ayuda de una la función `mean()` se pude calcular el promedio de una columna del conjunto de datos.

In [ ]:
df_clean['price'].mean()

3268.857414965986

De acuerdo al resultado de `mean()`, el costo promedio de una propiedad en el conjunto de datos es de **$3,268.87** dólares. 

**Pregunta 2. ¿Cuál es la ciudad que tienen el costo promedio más caro?**



La respuesta se obtiene mediante la agrupación con `groupby` de la columna `city` y la función `mean()`aplicada sobre la columna `price`.


In [ ]:
df_clean.groupby(['state','city'])['price'].mean().sort_values(ascending=False)

state  city             
CA     Pacific Palisades    12099.000000
NY     New York              5259.896552
MA     Roxbury               4895.833333
CA     Venice                4495.000000
       Hollywood             4187.400000
                                ...     
MA     Mattapan              2000.000000
NY     Forest Hills          1995.000000
       Jackson Heights       1899.000000
       Far Rockaway          1849.333333
CA     Sunland               1625.000000
Name: price, Length: 66, dtype: float64

De acuerdo con el resultado de la línea anterior, la ciudad de [`Pacific Palisades, CA`](https://www.google.com.mx/maps/place/Pacific+Palisades,+Los+%C3%81ngeles,+California,+EE.+UU./@34.0531621,-118.5683904,14z/data=!4m6!3m5!1s0x80c2a37a08897a89:0x4f7c6b3e45e83223!8m2!3d34.0356331!4d-118.5155901!16zL20vMDI3bDRx) tienen un costo promedio de ***`$12099.00`/montly***, mayor que el resto de las ciudades. Muy por arriba de la ciudad de Nueva York.
***(Lo valores estan expresados en dólares estadounidenses renta mensual***

**Pregunta 4. ¿Cuales son los vecindarios más costosos en promedios y en que ciudad se encuentran?**

In [ ]:
df_clean.groupby(['state','city','zipcode'])['price'].mean().sort_values(ascending=False)

state  city               zipcode
CA     Los Angeles        90067      18800.000000
NY     New York           10022      12146.133333
CA     Pacific Palisades  90272      12099.000000
NY     New York           10023       9510.111111
CA     Los Angeles        90049       6735.461538
                                         ...     
                          90032       1645.000000
       Sunland            91040       1625.000000
DC     Washington         20020       1610.636364
NY     Bronx              10472       1425.000000
CA     Los Angeles        90018       1033.000000
Name: price, Length: 281, dtype: float64

Entre los veindarios más costosos en primer lugar encontramos [**Century City**](https://www.google.com.mx/maps/place/Century+City,+California+90067,+EE.+UU./@34.0565614,-118.4181749,16z/data=!4m6!3m5!1s0x80c2bbf2ed672da7:0x4bf91367c431a816!8m2!3d34.0571327!4d-118.4148498!16s%2Fm%2F01zh8td), en **Los Angeles** con un costo promedio de **\$18,800.00**.

<p>(Alguien informado dice es **Hollywood**,las peliculas, etc; cosa de confirmar este último dato.)
<p>

Seguido por [**Nueva York, 10022**](https://www.google.com.mx/maps/place/Nueva+York,+10022,+EE.+UU./@40.7594581,-73.9694822,16.87z/data=!4m6!3m5!1s0x89c258e415b9da19:0x3e3afecde594d34a!8m2!3d40.7593941!4d-73.9697795!16s%2Fm%2F020cmyr) en en la ciudad de **Nueva York** con un costo promedio de **\$12,146.00**. 

In [ ]:
#Analizando la ciudad por su moda
print("\nModa por estado 'Estado'\n")
print("Moda 1: " ,df_clean.groupby('state')['city'].agg(pd.Series.mode))


Moda por estado 'Estado'

Moda 1:  state
CA    San Francisco
DC       Washington
MA           Boston
NY         New York
Name: city, dtype: object


In [ ]:
#Analizando la ciudad y codigo postal por su moda
print("\nModa por ciudad y codigo postal 'Estado'\n")
print("\nModa 2: " , df_clean.groupby('state')['city','zipcode'].agg(pd.Series.mode))


Moda por ciudad y codigo postal 'Estado'


Moda 2:                  city         zipcode
state                               
CA     San Francisco           94109
DC        Washington           20002
MA            Boston            2134
NY          New York  [11201, 11226]


<ipython-input-73-ed9f01325e4d>:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  print("\nModa 2: " , df_clean.groupby('state')['city','zipcode'].agg(pd.Series.mode))


In [ ]:
#Analizando el precio
print("\nMedia, Mediana y DEvStd del precio por: 'Estado'\n")
print(df_clean.groupby('state')['price'].agg(['mean', 'median', 'std']))

print("\nMedia, Mediana y DEvStd del precio por: 'Estado - Ciudad'\n")
print(df_clean.groupby(['state','city'])['price'].agg(['mean', 'median', 'std']))

print("\nMedia, Mediana y DEvStd del precio por: 'Estado - Ciudad-Codigo Postal' \n")
print(df_clean.groupby(['state','city','zipcode'])['price'].agg(['mean', 'median', 'std']))


Media, Mediana y DEvStd del precio por: 'Estado'

              mean  median          std
state                                  
CA     3360.989286  2974.5  1911.232249
DC     2574.139169  2219.0  1175.580103
MA     3453.485450  3145.0  1718.142311
NY     3826.867834  3100.0  3306.852501

Media, Mediana y DEvStd del precio por: 'Estado - Ciudad'

                                  mean  median          std
state city                                                 
CA    Canoga Park          2714.464286  2694.0   593.142446
      Chatsworth           2198.333333  2300.0   363.329511
      Culver City          3754.000000  3399.0  1131.407530
      Encino               2523.846154  2395.0   663.625126
      Hollywood            4187.400000  3662.0  1178.614780
...                                ...     ...          ...
NY    Rego Park            2400.000000  2400.0   636.396103
      Ridgewood            2350.000000  2350.0    70.710678
      Springfield Gardens  2295.000000  2295.0   